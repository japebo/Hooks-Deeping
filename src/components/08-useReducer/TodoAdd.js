import React from 'react'
import { useForm } from '../../hooks/useForm';


export const TodoAdd = ({handleAdd}) => {
    const [ formValue, handleInputChange, reset ] = useForm({
        task: ''
    });

    const handleSubmit = (e) => {
        e.preventDefault();
        
        if ( formValue.task.trim().length <= 1 ) {
            reset();
            return;
        }
        
        handleAdd( {
            id: new Date().getTime(),
            task: formValue.task,
            done: false    
        })

        reset();
    }

    return (
        <>
            <h4> Add Task </h4>

            <form onSubmit={ handleSubmit }>
                <input
                    type="text"
                    name="task"
                    className='form-control'
                    placeholder="Aprender..."
                    value={ formValue.task }
                    autoComplete="off"
                    onChange={ handleInputChange }
                />
                <button
                    type='submit'
                    className='btn btn-outline-primary mt-1 col-12'
                >
                    Add
                </button>
            </form>
        </>
    )
}
