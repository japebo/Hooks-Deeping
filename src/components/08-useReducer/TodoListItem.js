import React from 'react'

export const TodoListItem = ({todo, index, handleDeletion, handleToggle}) => {
    return (
        <li
            key={todo.id}
            className='list-group-item'
        >
            <p 
                className={ `${todo.done===true && 'complete'}` }
                onClick={ () => {handleToggle(todo.id)} } > {index+1}. {todo.task} </p>
            <button 
                className='btn btn-danger'
                onClick={ () => { handleDeletion(todo.id) } }
            >
                Delete
            </button>
        </li>
    )
}
