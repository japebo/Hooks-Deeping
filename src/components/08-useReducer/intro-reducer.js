
const initialState = [{
    id: 1,
    task: 'Botar la basura',
    done: false
}];

const todoReducer = ( state = initialState, action) => {

    if (action?.type === 'add') {
        return [...state, action.payload]
    }

    return state;
}

let todos = todoReducer();

const newTask = {
    id: 2,
    task: 'Buscar el pan',
    done: false
};

const actionToPerform = {
    type: 'add',
    payload: newTask
};

todos = todoReducer( todos, actionToPerform );

console.log(todos);