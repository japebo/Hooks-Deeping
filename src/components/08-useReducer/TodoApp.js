import React, { useEffect, useReducer } from 'react'
import './styles.css'
import { TodoAdd } from './TodoAdd';
import { TodoList } from './TodoList';
import { todoReducer } from './todoReducer';

const init = () => {
    return JSON.parse(localStorage.getItem('todos')) || [];

    // return [{
    //     id: new Date().getTime(),
    //     task: 'Aprender React',
    //     done: false
    // }];
}

export const TodoApp = () => {

    const [todos, dispatch] = useReducer(todoReducer, [], init);

    useEffect(() => {
        localStorage.setItem( 'todos', JSON.stringify(todos));
    }, [todos])


    const handleAdd = (newTodo) => {
        dispatch({
            type: 'add',
            payload: newTodo
        });
    }

    const handleDeletion = ( todoID ) => {
        const action = {
            type: 'delete',
            payload: todoID
        }

        dispatch(action);

    }

    const handleToggle = ( todoID ) => {

        dispatch({
            type: 'toggle',
            payload: todoID
        });

    }

    return (
        <div>
            <h1>ToDoApp ( {todos.length} )</h1>
            <hr/>

            <div className='row'>
                <div className='col-7'>
                    < TodoList todos={todos} handleDeletion={handleDeletion} handleToggle={handleToggle}/>
                </div>

                <div className='col-5'>
                    <TodoAdd handleAdd={handleAdd}/>
                </div>
            </div>
        </div>
    )
}
