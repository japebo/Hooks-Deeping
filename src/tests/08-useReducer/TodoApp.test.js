import { act } from "@testing-library/react";
import { mount, shallow } from "enzyme"
import { TodoApp } from "../../components/08-useReducer/TodoApp";
import { demoTodos } from "../fixtures/demoTodos";

describe('Pruebas en <TodoApp />', () => {
    const wrapper = shallow(<TodoApp />);
    
    test('Debe mostrarse correctamente', () => {
        expect(wrapper).toMatchSnapshot();
    })
    
    test('Debe agregar un TODO al LocalStorage', () => {
        const wrapper = mount(<TodoApp/>);

        Storage.prototype.setItem = jest.fn();

        act( () => {
            wrapper.find('TodoAdd').prop('handleAdd')(demoTodos[0]);
            wrapper.find('TodoAdd').prop('handleAdd')(demoTodos[1]);
        });

        expect(wrapper.find('h1').text().trim()).toBe('ToDoApp ( 2 )');
        expect( localStorage.setItem ).toHaveBeenCalledTimes(1);
    })
    
    test('Debe eliminar un TODO', () => {

        wrapper.find('TodoAdd').prop('handleAdd')(demoTodos[0]);
        expect(wrapper.find('h1').text().trim()).toBe('ToDoApp ( 1 )');
        wrapper.find('TodoList').prop('handleDeletion')(demoTodos[0].id);
        expect(wrapper.find('h1').text().trim()).toBe('ToDoApp ( 0 )');

    })
    
    
})
