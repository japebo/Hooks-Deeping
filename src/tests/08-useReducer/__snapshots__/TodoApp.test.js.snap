// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <TodoApp /> Debe mostrarse correctamente 1`] = `
<div>
  <h1>
    ToDoApp ( 
    0
     )
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-7"
    >
      <TodoList
        handleDeletion={[Function]}
        handleToggle={[Function]}
        todos={Array []}
      />
    </div>
    <div
      className="col-5"
    >
      <TodoAdd
        handleAdd={[Function]}
      />
    </div>
  </div>
</div>
`;
